ssl/ssl_lib.c

```diff
@@ -5073,3 +5100,142 @@ uint32_t SSL_get_max_early_data(const SSL *s)
+
+    lh_SSL_SESSION_free(ret->sessions);
+    OPENSSL_free(ret);
+    return NULL;
+}
+int SSL_SESSION_CACHE_up_ref(SSL_SESSION_CACHE *l)
+{
+    int i;
+
+    if (CRYPTO_atomic_add(&l->references, 1, &i, l->lock) <= 0)
+        return 0;
+
+    REF_PRINT_COUNT("SSL_SESSION_CACHE", l);
+    REF_ASSERT_ISNT(i < 2);
+    return ((i > 1) ? 1 : 0);
+}
+
+/* returns number of references, so 0 = freed */
+int SSL_SESSION_CACHE_free(SSL_SESSION_CACHE *c, SSL_CTX *ctx)
+{
+    int i;
+    SSL_SESSION *p, *next, *sentinel;
+
+    if (c == NULL)
+        return -1;
+
+    if (CRYPTO_atomic_add(&c->references, -1, &i, c->lock) <= 0)
+        return -1;
+
+    if (i != 0)
+        return i;
+
+    /*
+     * Both the linked list and hash are going away, so we can leave some
+     * unsanitized pointers hanging around for efficiency.
+     * We hold the only reference, and thus can forgo locking as well
+     * (but don't, for purity).
+     * All that needs to happen is to free each SSL_SESSION and ensure it's
+     * not in the hash table any more, and we NULL out the next/prev pointers
+     * just in case someone still holds a reference on the SSL_SESSION but
+     * not the list it was formerly on.
+     */
+    CRYPTO_THREAD_write_lock(c->lock);
+    sentinel = (SSL_SESSION *)&c->tail;
+    next = c->head;
+    p = NULL;
+    while (next != sentinel && next != NULL) {
+        p = next;
+        lh_SSL_SESSION_delete(c->sessions, p);
+        next = p->next;
+        p->not_resumable = 1;
+        p->next = p->prev = NULL;
+        if (ctx != NULL && c->remove_session_cb != NULL)
+            c->remove_session_cb(ctx, p);
+        SSL_SESSION_free(p);
+    }
+    lh_SSL_SESSION_free(c->sessions);
+
+    CRYPTO_THREAD_unlock(c->lock);
+    CRYPTO_THREAD_lock_free(c->lock);
+    OPENSSL_free(c);
+    return 0;
+}
+
+/*
+ * Use |key| as the query parameter to retrieve a matching SSL_SESSION from
+ * the (shared) session cache attached to |ctx|, incrementing the reference
+ * count on the returned pointer, and performing the necessary locking.
+ * The |ctx| lock must already be held (read or write).
+ */
+SSL_SESSION *SSL_CTX_get1_session(SSL_CTX *ctx, SSL_SESSION *key)
+{
+    SSL_SESSION_CACHE *c = ctx->session_cache;
+    SSL_SESSION *ret;
+
+    /*
+     * There is an interesting philosophical argument here, in that retrieving
+     * from an LHASH actually performs a write(!), to the stats counters.
+     * But, we disregard that race and can accept some invalid stats.
+     */
+    CRYPTO_THREAD_read_lock(c->lock);
+    ret = lh_SSL_SESSION_retrieve(c->sessions, key);
+    if (ret != NULL)
+        SSL_SESSION_up_ref(ret);
+    CRYPTO_THREAD_unlock(c->lock);
+    return ret;
+}
+
+/* Makes 'b' use 'a's session cache */
+int SSL_CTX_share_session_cache(SSL_CTX *a, SSL_CTX *b)
+{
+    int ret = 0;
+    SSL_SESSION_CACHE *save;
+
+    if (a == b)
+        return 0;
+
+    CRYPTO_THREAD_write_lock(b->lock);
+    CRYPTO_THREAD_read_lock(a->lock);
```

Comment:
Yup, although I would except that SSL_CTX setup occur in an initialization thread, and that two threads wouldn't be modifying separate SSL_CTX's. Although, I did have a design where there was one SSL_CTX per thread that was affinity locked to a processor core... not that we used that.